ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala

```diff
@@ -0,0 +1,138 @@
+    *  This implementation ensures that all the time the "smallest" available TASK will be picked for work.
+    *
+    * Please note:
+    *  - If workerParallelism is one, this should result in monotonously increasing execution sequence
+    *    (regardless of the demand downstream)
+    *  - If workerParallelism is equal or bigger than the number of initial n tasks, and downstream is faster,
+    *    then prioritization has on time to kick in (the backing priority queue will have mostly one element), so
+    *    execution order will be similar to simple parallel execution of sequences of tasks
+    *
+    * @param workerParallelism defines the maximum parallelism of unordered processing.
+    *                          Naturally capped by size of initialTasks
+    * @param work The worker function, asynchronous computation should return a RESULT,
+    *             and the next TASK, or no TASK if TASK processing is finished
+    * @param initialTasks The collection of initial TASKS for execution
+    * @tparam TASK type of TASKs, needs to have an Ordering defined.
+    *              Always the "smallest" task will be selected for execution
+    * @tparam RESULT The ype of the RESULT
+    * @return A Source, with TASK, RESULT pairs.
+    *         The ordering of the elements will simply follow the work completion order.
+    *         Completes, if all TASKS finish (for all of them a final work executed, giving no continuation)
+    *         Fails if work processing fails.
+    */
+  def pullWorkerSource[TASK: Ordering, RESULT](
+      workerParallelism: Int,
+      materializer: Materializer,
+  )(
+      work: TASK => Future[(RESULT, Option[TASK])],
+      initialTasks: Iterable[TASK],
+  ): Source[(TASK, RESULT), NotUsed] = if (initialTasks.isEmpty) Source.empty
+  else {
+
+    val (signalQueue, signalSource) = Source
+      .queue[Unit](initialTasks.size)
+      .preMaterialize()(materializer)
+
+    val queueState = new QueueState(signalQueue, initialTasks)
+
+    signalSource
+      .mapAsyncUnordered(workerParallelism) { _ =>
+        val task = queueState.startTask
+        work(task).map { case (result, nextTask) =>
+          queueState.finishTask(nextTask)
+          task -> result
+        }(DirectExecutionContext)
+      }
+  }
+
+  /** Helper class to capture stateful  operations of pullWorkerSource
+    */
+  class QueueState[TASK: Ordering](
+      signalQueue: BoundedSourceQueue[Unit],
+      initialTasks: Iterable[TASK],
+  ) {
+    private val priorityQueue =
+      new mutable.PriorityQueue[TASK]()(implicitly[Ordering[TASK]].reverse)
+    private var runningTasks: Int = 0
+
+    initialTasks.foreach(addTask)
+
+    def startTask: TASK = synchronized {
+      runningTasks += 1
+      priorityQueue.dequeue()
+    }
+
+    def finishTask(nextTask: Option[TASK]): Unit = synchronized {
+      nextTask match {
+        case None if priorityQueue.isEmpty && runningTasks == 1 =>
```

Comment:
I think there is a problem with the runningTasks counter. You have not run into it yet, because unit tests only use parallelism of 1.
Imagine there are 2 tasks without continuation and the parallelism is 2. They both will go through `case None` without decrement of runningTasks. As a result the signalQueue will never complete.